WebDriver: [macOS] Add support for Wheel input source and actions

patrickangle opened the pull request:
#### 26e467ec43ea1c61b517fe20b4bb2d7ad7801a6d
<pre>
WebDriver: [macOS] Add support for Wheel input source and actions
<a href="https://bugs.webkit.org/show_bug.cgi?id=244092">https://bugs.webkit.org/show_bug.cgi?id=244092</a>
rdar://45649400

Reviewed by Devin Rousso.

Enable WEBDRIVER_WHEEL_INTERACTIONS on macOS and properly create and dispatch scroll events. Because NSEvent does not
provide a direct mechanism for creating synthetic scroll events (like it does for Mouse and Keyboard events) we create a
CGEvent to represent the scroll, create an NSEvent from the CGEvent, and then force the event to be relative to and
associated with the window, which makes the events consistent with our other window-relative events and prevents the
event from being thrown out by the window as irrelevant.

This passes the WPT tests in /webdriver/tests/perform_actions/wheel.py when paired with a matching safaridriver change
to support parsing the events.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebKit/UIProcess/Automation/mac/WebAutomationSessionMac.mm:
(WebKit::viewportLocationToWindowLocation):
(WebKit::WebAutomationSession::platformSimulateMouseInteraction):
(WebKit::WebAutomationSession::platformSimulateWheelInteraction):
* Source/cmake/OptionsMac.cmake:

Canonical link: <a href="https://commits.webkit.org/253578@main">https://commits.webkit.org/253578@main</a>
</pre>














<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d13ee6b0b67e384f473034af3d7de176c5beecc7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/86431 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/30351 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/17401 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/95279 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/148987 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/90419 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/28751 "Built successfully") | [✅ ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/25337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/78584 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/90522 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92041 "Passed tests") | [✅ ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/23324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/73405 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23381 "Passed tests") | 
| | [✅ ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/78341 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/78730 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66385 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/78368 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/26664 "Built successfully") | [✅ ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/12580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/72003 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/26578 "Built successfully") | [✅ ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/13592 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/25738 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2547 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28257 "Built successfully") | [✅ ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/73405 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/74784 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28197 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/32875 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16535 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebKit/UIProcess/Automation/mac/WebAutomationSessionMac.mm:
```diff
@@ -789,6 +799,29 @@ static BOOL characterIsProducedUsingShift(NSString *characters)
+
+void WebAutomationSession::platformSimulateWheelInteraction(WebPageProxy& page, const WebCore::IntPoint& locationInViewport, const WebCore::IntSize& delta)
+{
+    static constexpr auto scrollWheelCount = 2;
```
Why `2`?

patrickangle replied:
From the documentation, this is the number of "wheels" on the mouse: https://developer.apple.com/documentation/coregraphics/1541327-cgeventcreatescrollwheelevent

In practice, each wheel represents a single axis, starting with the Vertical axis scroll wheel (y), followed by the horizontal axis scroll wheel (x), so we have two wheels in order to represent the different X and Y delta.

dcrousso commented on Source/WebKit/UIProcess/Automation/mac/WebAutomationSessionMac.mm:
```diff
@@ -789,6 +799,29 @@ static BOOL characterIsProducedUsingShift(NSString *characters)
+    // Set the CGEvent location in flipped coords relative to the first screen, which compensates for the behavior of
+    // +[NSEvent eventWithCGEvent:] when the event has no associated window. See <rdar://problem/17180591>.
+    auto locationOnScreen = [window convertPointToScreen:locationInWindow];
+    locationOnScreen = CGPointMake(locationOnScreen.x, NSScreen.screens.firstObject.frame.size.height - locationOnScreen.y);
```
Should we use something like `window.screen.frame.size.height` instead?

patrickangle replied:
I'll need to do some more testing here to determine if that works. I realize now that both my monitors are the exact same resolution, so this will always be right on my machine currently 😅

patrickangle replied:
I just double checked with two different screen resolutions and some silly screen arrangements and this does in fact need to be the height of the **first** screen, not the screen containing our eventual target window.

patrickangle committed:
WebDriver: [macOS] Add support for Wheel input source and actions
https://bugs.webkit.org/show_bug.cgi?id=244092
rdar://45649400

Reviewed by Devin Rousso.

Enable WEBDRIVER_WHEEL_INTERACTIONS on macOS and properly create and dispatch scroll events. Because NSEvent does not
provide a direct mechanism for creating synthetic scroll events (like it does for Mouse and Keyboard events) we create a
CGEvent to represent the scroll, create an NSEvent from the CGEvent, and then force the event to be relative to and
associated with the window, which makes the events consistent with our other window-relative events and prevents the
event from being thrown out by the window as irrelevant.

This passes the WPT tests in /webdriver/tests/perform_actions/wheel.py when paired with a matching safaridriver change
to support parsing the events.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebKit/UIProcess/Automation/mac/WebAutomationSessionMac.mm:
(WebKit::viewportLocationToWindowLocation):
(WebKit::WebAutomationSession::platformSimulateMouseInteraction):
(WebKit::WebAutomationSession::platformSimulateWheelInteraction):
* Source/cmake/OptionsMac.cmake:

Canonical link: https://commits.webkit.org/253578@main
